Plugins: Display eligibility warnings for dotcom sites

vindl opened the pull request:
**Update**: it seems that after the rebase https://github.com/Automattic/wp-calypso/pull/9880 is making it impossible to get to the warnings page from plugin detail page. If you want to test the component you can navigate directly to: `plugins/bbpress/{siteSlug}/eligibility`

**Testing instructions**:
1. Dispatch this action from your console:

```javascript
dispatch( { 
    type: actionTypes.AUTOMATED_TRANSFER_ELIGIBILITY_REQUEST,
    siteId: '{siteId}',
} );
```

2. navigate to `plugins/bbpress/{siteSlug}/eligibility` (due to bug mentioned above, getting here from by clicking on `Install` button wont work atm)
3. Verify that correct errors and warnings are shown for you test site.




**Visual changes:**
* Plugin detail page for dotcom site:

![step1](https://cloud.githubusercontent.com/assets/1182160/21201170/34060c92-c24a-11e6-99b4-6f5679a76992.png)

* Eligibility warnings component:

![step2](https://cloud.githubusercontent.com/assets/1182160/21201174/3995fc6c-c24a-11e6-82fe-70cffd4f8d28.png)

dmsnell commented on client/my-sites/plugins-wpcom/plugin-install-button.jsx:
```diff
@@ -9,12 +9,17 @@ import { localize } from 'i18n-calypso';
 		primary={ true }
 		type="submit"
 		disabled={ disabled }
+		href={ getPluginLink( plugin.slug, site.slug ) }
```
this is probably simple enough to simply inline it without the function call

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+			<div className="eligibility-warnings">
+				<SectionHeader label={ translate( 'Conflicts' ) } />
+
+				{ errors.map( ( message ) =>
```
This could be a good place for destructuring

```js
{ errors.map( ( { actionUrl, description, title } ) => (
	…
) ) }
```

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

roundhill commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+							'You must resolve the errors above before proceeding. '
+						) }
+						{ isEligible && translate(
+							'By proceeding you understand that you will loose these features. '
```
What if we said it more like:

`If you proceed you will no longer be able to use these features.`

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+						<Gridicon icon="notice" className="eligibility-warnings__error-icon" />
+						<div className="eligibility-warnings__message-content">
+							<div className="eligibility-warnings__message-title">
+								{ translate( 'Error' ) + `: ${ message.title }` }
```
we should probably include this into the translation since it could be handled differently, especially with some languages that are RTL

```js
{ translate( 'Error: %( title )', { args: { title } } }
```

vindl replied:
Good point, added in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2. Just noting that it won't work with spaces, must be `%(title)s`.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+					</Card>
+				) }
+
+				{ pluginWarnings.map( ( message ) =>
```
See destructuring comment above

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+						<Gridicon icon="notice" className="eligibility-warnings__warning-icon" />
+						<div className="eligibility-warnings__message-content">
+							<div className="eligibility-warnings__message-title">
+								{ translate( 'Unsupported feature' ) + `: ${ message.title }` }
```
See translation comment about

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+					</Card>
+				) }
+
+				{ widgetWarnings.map( ( message ) =>
```
See destructuring comment above

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+						<Gridicon icon="notice" className="eligibility-warnings__warning-icon" />
+						<div className="eligibility-warnings__message-content">
+							<div className="eligibility-warnings__message-title">
+								{ translate( 'Incompatible widget' ) + `: ${ message.title }` }
```
See translation comment above

vindl replied:
Updated in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+						{ ! isEligible && translate(
+							'You must resolve the errors above before proceeding. '
+						) }
+						{ isEligible && translate(
```
We may have no warnings too. In this case we wouldn't even want to show this dialog I think. @drw158 can clarify but he made a nice diagram showing the flow here.

drw158 replied:
![install plugin flow](https://cloud.githubusercontent.com/assets/618551/21230323/39cf051a-c2aa-11e6-9de3-7b34f1f9cd4c.png)

roundhill commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+						{ isEligible && translate(
+							'By proceeding you understand that you will loose these features. '
+						) }
+						{ translate( 'Have questions? Please contact support.' ) }
```
It'd be good to link to support here: https://wordpress.com/help/contact

drw158 replied:
Specifically `contact support` should be linked.

vindl replied:
Added link to support in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

dmsnell commented on client/my-sites/plugins/plugin-eligibility/index.jsx:
```diff
@@ -0,0 +1,85 @@
+	render() {
+		const { translate } = this.props;
+
+		// Mock some errors and warnings until we get real data from the API
```
once we get #9914 merged, we can create a snippet in Chrome's dev tools with this state and directly run that from the console. that way we can easily fake the state without even polling or mocking it

```js
dispatch( {
	type: actionTypes.AUTOMATED_TRANSFER_ELIGIBILITY_UPDATE,
	status: 'conflicts',
	lastUpdate: Date.now(),
	eligibilityHolds: [ { … }, { … }, { … } ],
	eligibilityWarnings: [ … ]
} );
```

roundhill replied:
> that way we can easily fake the state without even polling or mocking it

👍

vindl replied:
I think it will be easier to just use `AUTOMATED_TRANSFER_ELIGIBILITY_REQUEST` to fetch real data from the API. I added that in testing instruction.

drw158 commented on client/my-sites/plugins/eligibility-warnings/style.scss:
```diff
@@ -0,0 +1,68 @@
+    display: flex;
+    align-items: center;
+    flex-direction: row;
+    height: 80px;
```
This height is causing some mobile layout problems. Getting rid of this line should work.

drw158 commented on client/my-sites/plugins/eligibility-warnings/style.scss:
```diff
@@ -0,0 +1,68 @@
+    margin: 0 16px;
+}
+
+.eligibility-warnings__message-description {
```
Might as well combine these 2 style blocks, since they are both `color: $gray;`

folletto commented:
> I think a combination of flex-wrap and flex-basis should work. When it gets squished too much, the buttons should drop below the text.

Agreed. :)

seear commented:
The _Conflicts_ card below can be used by the [theme upload](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/themes/theme-upload/index.jsx) page:
![step2](https://cloud.githubusercontent.com/assets/1182160/21201174/3995fc6c-c24a-11e6-82fe-70cffd4f8d28.png)

Can we move it to a shared area such as `/client/components/`?

vindl commented on client/my-sites/plugins/eligibility-warnings/index.jsx:
```diff
@@ -0,0 +1,142 @@
+import Gridicon from 'components/gridicon';
+import SectionHeader from 'components/section-header';
+
+class EligibilityWarnings extends Component {
```
Refactored as stateless functional component in 8b3ab4d82f850f5e3baa9ee0a73ef4995179f0d2.

vindl commented:
@drw158 Thanks for catching that! I updated it according to your suggestions and it should be working as expected now.

@seear I was thinking about that too, but wasn't sure if `client/components` was the right place. I've moved it there for now.

seear commented:
>  I was thinking about that too, but wasn't sure if client/components was the right place. I've moved it there for now.

Thanks @vindl. Thinking about it a bit more, `/client/blocks` might be a better place, since I think `/client/components` is generally for primitive UI parts, whereas we have something slightly more complex here, for a specific purpose.

seear commented:
> Thanks @vindl. Thinking about it a bit more, /client/blocks might be a better place, since I think /client/components is generally for primitive UI parts, whereas we have something slightly more complex here, for a specific purpose.

(I'm happy to move it later in a different PR if you want to get this merged quickly)

dmsnell commented on client/my-sites/plugins/index.js:
```diff
@@ -76,6 +76,14 @@ module.exports = function() {
 		);
 
+		if ( config.isEnabled( 'automated-transfer' ) ) {
+			page( '/plugins/:plugin/:site_id/eligibility',
```
I'm pausing here to reflect on the URL structure:

 - do we want to expose `eligibility` in the URL? probably?
 - do we want to tie this to `/plugins`? maybe
 - do we want it instead to be at `/transfers/eligibility`? maybe not?

this would be a good thing for a design review.

vindl replied:
We agreed to postpone this for future PRs, until we decide on appropriate path scheme.

dmsnell commented on client/my-sites/plugins/plugin-eligibility/index.jsx:
```diff
@@ -0,0 +1,85 @@
+		return `/plugins/${ pluginSlug }/${ siteSlug }`;
+	};
+
+	goBack = () => page( this.getBackUrl() );
```
could we pull this outside of the component and pass it in? It would be far easier to test the component for one if we didn't directly call page. here is _one way_ we could do this:

```js
this.props.navigateTo( this.getBackUrl() );

const withNagivation = Component => props => <Component { ...{ props }, navigateTo: page } />

export default withNavigation( localize( PluginEligibility ) );
```

vindl commented:
@seear I'm happier with `client/blocks` too, moved it there already. 
(btw. I've encountered a CSS issue after moving it there that broke the card layout; applied a band-aid fix for now)

@dmsnell @roundhill I would prefer if we could resolve other concerns in followup PRs, so we can work on different things in parallel and hopefully get them done more quickly. Let me know what do you think?

dmsnell commented:
> I would prefer if we could resolve other concerns in followup PRs

which concerns? let's try to not push something into **master** that we know shouldn't belong. if the concerns are ancillary to this task or if they are refactorings of other existing code I'm fine pushing them off, but I'm less excited about introducing new code we know we don't want.

vindl commented:
@dmsnell For example: https://github.com/Automattic/wp-calypso/pull/10052#discussion_r92671055, https://github.com/Automattic/wp-calypso/pull/10052#pullrequestreview-13193305.
I thought that we were aiming on releasing a quick prototype that calls `initiate` transfer until Friday for internal testing, and that we will iterate on this after that anyway. But if you all don't agree, I don't mind adding more stuff in this PR.

dmsnell commented:
> I thought that we were aiming on releasing a quick prototype that calls initiate transfer until Friday for internal testing, and that we will iterate on this after that anyway. But if you all don't agree, I don't mind adding more stuff in this PR.

yeah @vindl my "pause" was just that - a recorded note to think about this thing for later

on the `page` thing I think there are quick and easy ways around building it into the component itself and I'd prefer to keep that out of **master** rather than "fix it later"

vindl commented:
@dmsnell I think I've addressed all of the feedback, expect for this one https://github.com/Automattic/wp-calypso/pull/10052/files#r92671055 until we decide what is the best way to organize those paths.

dmsnell commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -17,66 +20,90 @@ const EligibilityWarnings = props => {
+			description: translate( 'This feature is not supported on private sites.' ),
+			supportUrl: 'https://support.wordpress.com/'
+		}
+	};
```
This is fine, but feel free to move this into a function in the module scope which accepts `translate` as an argument. That can clean up the component proper so less noise is in the render path.

dmsnell commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -117,29 +144,16 @@ const EligibilityWarnings = props => {
+};
+
+const mapStateToProps = state => {
+	const siteSlug = getSelectedSiteSlug( state );
```
Shall we keep to the `siteId` for now? I think that may be slightly more resilient across site kinds

dmsnell replied:
We also need to think about what happens when someone navigates to `All Sites` which makes the selected site null. Maybe this isn't a problem because of how the controller is setup. This is not my expertise.

vindl commented on client/my-sites/plugins/plugin-eligibility/index.jsx:
```diff
@@ -0,0 +1,85 @@
+				description: translate( 'Authors widget will no longer work after this plugin is installed.' ),
+				actionUrl: 'https://en.support.wordpress.com/'
+			}
+		];
```
Test data removed, using data from the state instead.

dmsnell replied:
test data still there?

dmsnell commented:
We manually request site eligibility, you can enter this in the developer console on a local build, replacing `1337` with whatever `siteId` you want.

```js
dispatch( {
    type: actionTypes.AUTOMATED_TRANSFER_ELIGIBILITY_REQUEST,
    siteId: 1337
} );
```

vindl commented:
> To manually request site eligibility you can enter this in the developer console on a local build, replacing 1337 with whatever siteId you want.

@dmsnell Updated testing instructions to account for this, along with the selector call.

dmsnell commented on client/my-sites/plugins/plugin-eligibility/index.jsx:
```diff
@@ -0,0 +1,54 @@
+	}
+}
+
+const withNavigation = WrappedComponent => props => <WrappedComponent { ...{ ...props, navigateTo: page } } />;
```
You should probably say something here like `@dmsnell made me do it` so people know not to yap at you

vindl replied:
Check out the comment and let me know what do you think. :smile:

dmsnell commented on client/state/automated-transfer/constants.js:
```diff
@@ -9,6 +9,8 @@ export const transferStates = {
 export const eligibilityHolds = {
 	MULTIPLE_USERS: 'MULTIPLE_USERS',
+	NO_BUSINESS_PLAN: 'NO_BUSINESS_PLAN',
+	NO_VIP_SITES: 'NO_VIP_SITES',
```
Can you add these to the API mapping as well?

`state/data-layer/wpcom/sites/automated-transfer/eligibility/index.js`

vindl replied:
Added.
